standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/properties/PropertyMap.java

```diff
@@ -0,0 +1,466 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.metastore.properties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.io.InvalidObjectException;
+import java.io.ObjectInputStream;
+import java.io.ObjectStreamException;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.TreeMap;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+/**
+ * A property map pertaining to a given object type (cluster, database, table).
+ * <p>
+ *   Maps follow a copy-on-write scheme gated by a dirty flag (avoid copy of a dirty map). This allows
+ *   sharing their content (the inner properties map) with guaranteed isolation and thread safety.
+ * </p>
+ */
+public class PropertyMap implements Serializable {
```

Comment:
nit: may be we can declare a super class for de/serializing instead extends `Serializable`, for example, the super class might look like:
```java
abstract class Abc {
  public Abc(DataInput input, Object... args) {
    // noop
  }
  public abstract Abc write(DataOutput out);
}

Reply:
We could but this is not needed and would hide the convention which does not require any extending any class, just implement the interface. And note that most serialization methods are private (
  private Object writeReplace() throws ObjectStreamException; private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException).

Reply:
I check the SerializationProxy, looks like we use something like `public Abc(DataInput input, Object... args)` to deserialize the object from byte array, and use `write(DataOutput out)` to serialize the instance, so my idea is that, we'd better to have these two methods as a template for SerializationProxy.

Reply:
The idea behind the serialization proxy is that of a convention; the 'contract' does not need to be made explicit through inheritance. Adding an abstract class for one use case that is already provably following a convention does not bring any clarity imho.